File: sdk/js/tests/unit/default-mutation.test.ts
import { describe, it, expect } from "vitest";
import { any, unknown, object } from "../../src/index.js";

// Repro for AVV-007: mutable default values shared between validations.
// Pass-through schemas (any/unknown) return the default value by reference,
// so mutating one parse result corrupts the default for the next parse.
describe("Mutable default isolation (AVV-007)", () => {
  it("any() array default is not shared between parses", () => {
    const s = any().default([] as unknown[]);
    const a = s.parse(undefined) as unknown[];
    a.push("mutated");
    const b = s.parse(undefined) as unknown[];
    expect(b).toEqual([]);
  });

  it("unknown() object default is not shared between parses", () => {
    const s = unknown().default({ tags: [] as string[] });
    const a = s.parse(undefined) as { tags: string[] };
    a.tags.push("x");
    const b = s.parse(undefined) as { tags: string[] };
    expect(b.tags).toEqual([]);
  });

  it("object field with any() default is not shared", () => {
    const s = object({ meta: any().default({ count: 0, items: [] as number[] }) });
    const a = s.parse({}) as { meta: { count: number; items: number[] } };
    a.meta.items.push(1);
    a.meta.count = 99;
    const b = s.parse({}) as { meta: { count: number; items: number[] } };
    expect(b.meta).toEqual({ count: 0, items: [] });
  });
});

File: sdk/go/default_mutation_test.go
package anyvali

import "testing"

// Repro for AVV-007: mutable default values shared between parses.
// The default value is stored by reference and handed to validation. For
// schemas that pass values through (Any/Unknown items, allowed unknown keys),
// the materialized result aliases the stored default, so mutating one parse
// result corrupts the default for the next parse.

func TestDefaultNotSharedArrayOfAny(t *testing.T) {
	s := Array(Any()).Default([]any{map[string]any{"n": float64(0)}})

	r1, err := s.Parse(nil)
	if err != nil {
		t.Fatalf("first parse failed: %v", err)
	}
	m1 := r1.([]any)[0].(map[string]any)
	m1["n"] = float64(99)

	r2, err := s.Parse(nil)
	if err != nil {
		t.Fatalf("second parse failed: %v", err)
	}
	m2 := r2.([]any)[0].(map[string]any)
	if m2["n"] != float64(0) {
		t.Errorf("default mutated across parses: got n=%v, want 0", m2["n"])
	}
}

func TestDefaultNotSharedObjectAllowedKey(t *testing.T) {
	s := Object(map[string]Schema{}).
		UnknownKeys(Allow).
		Default(map[string]any{"tags": []any{"orig"}})

	r1, err := s.Parse(nil)
	if err != nil {
		t.Fatalf("first parse failed: %v", err)
	}
	// Mutate the slice element in place: if the slice aliases the stored
	// default's backing array, this corrupts the default for later parses.
	r1.(map[string]any)["tags"].([]any)[0] = "MUTATED"

	r2, err := s.Parse(nil)
	if err != nil {
		t.Fatalf("second parse failed: %v", err)
	}
	tags2 := r2.(map[string]any)["tags"].([]any)
	if tags2[0] != "orig" {
		t.Errorf("default mutated across parses: got tags[0]=%v, want orig", tags2[0])
	}
}
